mirascope/tools/base.py

```diff
@@ -37,28 +39,75 @@ class ConfigurableTool(
     model_config = ConfigDict(arbitrary_types_allowed=True)
 
     @classmethod
-    def _config(cls) -> _ToolConfigT:
+    def _get_config(cls) -> _ToolConfigT:
         """Get tool configuration"""
         return cast(_ToolConfigT, cls.__config__)
 
     @classmethod
     def from_config(
-        cls, config: _ToolConfigT
-    ) -> type[ConfigurableTool[_ToolConfigT, _ToolSchemaT]]:
+        cls: type[_ConfigurableToolT], config: _ToolConfigT
+    ) -> type[_ConfigurableToolT]:
         """Create tool class with custom configuration"""
 
         # ClassVar cannot get TypeVar Type. So, we need the comment to ignore the error.
         config = cls.__config__.model_validate(config)  # pyright: ignore [reportAssignmentType]
-        NewModel = create_model(
+        new_model = create_model(
             cls.__name__,
             __base__=cls,
             __module__=cls.__module__,
         )
-        NewModel.__prompt_usage_description__ = cls.__prompt_usage_description__
-        NewModel.__config__ = config
-        return NewModel
+        new_model.__prompt_usage_description__ = cls.__prompt_usage_description__
+        new_model.__config__ = config
+        return new_model
 
     @classmethod
     def usage_description(cls) -> str:
         """Returns instructions for using this tool."""
         return inspect.cleandoc(cls.__prompt_usage_description__)
+
+
+_ConfigurableToolT = TypeVar("_ConfigurableToolT", bound=ConfigurableTool)
+
+
+class ConfigurableToolKit(BaseToolKit, Generic[_ToolConfigT], ABC):
+    """Abstract base class for configurable toolkit.
+
+    Subclasses must define a `__prompt_usage_description__` class variable
+    and __config__ class variable with a subclass of _ToolConfig.
+    """
+
+    __config__: ClassVar[_ToolConfig]
```

Comment:
hmm, realizing that `__config__` shouldn't be a `ClassVar` (both for `ConfigurableTool` and `ConfigurableToolKit`). The configuration is not something to be defined at the class level but something that should be defined for every instance (e.g. you can have two of the same tools or toolkits with different configurations, which doesn't really work well with class vars).

What if instead we did `_config: _ConfigurableToolConfigT`? We can still set the default variable, and we can still construct the class from config, but now it's clear we can have a different config for each tool. We could also do `__config__: PrivateAttr[_ConfigurableToolConfigT]`. Which do you prefer?

Reply:
I think instead what we should do is:
1. Define toolkit configuration as fields of the class
2. We should define a tool config for the individual tools inside of the toolkit
3. We should set the individual tool configs using the toolkit's fields on creation (i.e. `create_tools`)